### PR TITLE
[spec] Update spec for option B'

### PR DESCRIPTION
**Pull request by rossberg**
This updates the spec document to option B'.

@ioannad, would you be interested in reviewing this change? I know it is rather large, but it's deleting more than it's adding. :)

### CODE REVIEW

**Comment by conrad-watt** [comment]
@rossberg I can probably take a look at this while I'm in the air on Sunday.

**Review by titzer — APPROVED** [review]
I'd be OK merging this PR into `exnref-1b` before review and then reviewing the merged branch as a PR on the main branch. (AFAICT there is no PR to merge `exnref-1b` into the main branch yet, and I already noticed a typo there).

**Comment by rossberg** [comment]
> AFAICT there is no PR to merge exnref-1b into the main branch yet

There is: #285 ;)

Happy to do it in either order.

**Comment by rossberg** [comment]
Thinking again, it probably makes sense not to block #285, so people see an up-to-date explainer and tests etc. I rather hold off on merging this one until @conrad-watt or somebody else did a proper review.

**Comment by conrad-watt** [comment]
Apologies for the delay, on this now.

**Comment by conrad-watt** [comment]
EDIT: I originally complained about a build warning here, but I was building on the wrong branch

File: document/core/appendix/index-instructions.py
Comment (conrad-watt) [inline document/core/appendix/index-instructions.py:113]: Are we no longer listing the `catch` opcodes separately?

File: document/core/appendix/properties.rst
Comment (conrad-watt) [inline document/core/appendix/properties.rst:62]: It feels kind of nasty to classify an instruction inside a context as a "result". Can you walk me through why you don't want to use something like `\THROWadm` any more?

File: document/core/util/macros.def
Comment (conrad-watt) [inline document/core/util/macros.def:1290]: line 1283 (above): `.. |vdashval| mathdef:: \xref{appendix/properties}{valid-val}{\vdash}`

should be `exec/modules` rather than `appendix/properties` (I think)

File: document/core/exec/instructions.rst
Comment (conrad-watt) [inline document/core/exec/instructions.rst:2752]: Is it necessary to allocate exceptions in the store? Since they're immutable, can the abstract semantics not just treat them as ordinary values?

File: document/core/appendix/index-instructions.py
Comment (rossberg) [reply document/core/appendix/index-instructions.py:113]: They are no longer in the space of instruction opcodes, but just flag bytes in a vector.

File: document/core/appendix/properties.rst
Comment (rossberg) [reply document/core/appendix/properties.rst:62]: The introduction of the context is actually unrelated to replacing THROWadm, its absence was simply wrong before. We need it here since there is no reduction rule that can eliminate it on the outermost level (outside a function). We can't simply add a rule XT[exn] ⇝ exn, since that would introduce overlapping rules.

I see two ways to fix this in the future: (1) switching to a bubbling semantics for exceptions (avoiding overlap), or (2) introducing a separate global reduction relation like we have with threads. The former should go along with changes to other parts of the semantics to stay coherent, the latter I'd rather do after we have merged with threads. Hence I don't want to do either in this PR.

File: document/core/exec/instructions.rst
Comment (rossberg) [reply document/core/exec/instructions.rst:2752]: References need internal identity, because it can become observable in the host, e.g. in JS. This is analogous to functions, which are likewise immutable.

(The lack of exn identity was one of the problems we faced with the JS API spec in the old proposal.)

File: document/core/util/macros.def
Comment (rossberg) [reply document/core/util/macros.def:1290]: Indeed, fixed.

**Review by rossberg — COMMENTED** [review]
Thanks for the review!

**Review by conrad-watt — APPROVED** [review]
In the longer term, I'd like to revisit the way the exception "value" is formalised, but beyond that things seem fine.

**Pull request merged**
